t/cs.t

```diff
@@ -2191,3 +2191,302 @@ received: hello (text)
+                return ngx.exit(444)
+            end
+            ngx.log(ngx.INFO, string.format("host: <%s>", ngx.var.http_host))
+        }
+    }
+--- request
+GET /c
+--- error_log
+host: <client.test:8080>
+
+
+
+
+=== TEST 33: SNI derived from custom host header (without port number)
+--- http_config eval: $::HttpConfig
+--- config
+    listen 12345 ssl;
+    server_name test.com;
+    ssl_certificate ../../cert/test.crt;
+    ssl_certificate_key ../../cert/test.key;
+    server_tokens off;
+
+    location = /c {
+        content_by_lua_block {
+            local client = require "resty.websocket.client"
+            local wb, err = client:new()
+
+            local uri = "wss://127.0.0.1:12345/s"
+            local opts = {
+                host = "test.com",
+                ssl_verify = false,
+            }
+            local ok, err = wb:connect(uri, opts)
+            if not ok then
+                ngx.say("failed to connect: ", err)
+                return
+            end
+        }
+    }
+
+    location = /s {
+        content_by_lua_block {
+            local server = require "resty.websocket.server"
+            local wb, err = server:new()
+            if not wb then
+                ngx.log(ngx.ERR, "failed to new websocket: ", err)
+                return ngx.exit(444)
+            end
+
+            ngx.log(ngx.INFO, string.format("host: <%s>", ngx.var.http_host))
+            ngx.log(ngx.INFO, "SSL server name: ", ngx.var.ssl_server_name)
+        }
+    }
+--- request
+GET /c
+--- error_log
+host: <test.com>
+SSL server name: test.com
+
+
+
+=== TEST 34: SNI derived from custom host header (with port number)
+--- http_config eval: $::HttpConfig
+--- config
+    listen 12345 ssl;
+    server_name test.com;
+    ssl_certificate ../../cert/test.crt;
+    ssl_certificate_key ../../cert/test.key;
+    server_tokens off;
+
+    location = /c {
+        content_by_lua_block {
+            local client = require "resty.websocket.client"
+            local wb, err = client:new()
+
+            local uri = "wss://127.0.0.1:12345/s"
+            local opts = {
+                host = "test.com:8443",
+                ssl_verify = false,
+            }
+            local ok, err = wb:connect(uri, opts)
+            if not ok then
+                ngx.say("failed to connect: ", err)
+                return
+            end
+        }
+    }
+
+    location = /s {
+        content_by_lua_block {
+            local server = require "resty.websocket.server"
+            local wb, err = server:new()
+            if not wb then
+                ngx.log(ngx.ERR, "failed to new websocket: ", err)
+                return ngx.exit(444)
+            end
+
+            ngx.log(ngx.INFO, string.format("host: <%s>", ngx.var.http_host))
+            ngx.log(ngx.INFO, "SSL server name: ", ngx.var.ssl_server_name)
+        }
+    }
+--- request
+GET /c
+--- error_log
+host: <test.com:8443>
+SSL server name: test.com:8443
```

Comment:
It may be so that the protocol allows for any string of characters to be an SNI, but proper SNI usage while establishing a TLS connection dictates that the SNI in that case be the name of the virtual host, i.e. the DNS hostname; so, no port by default. If the user overrides with `server_name`, then anything goes.

Reply:
Done. Not sure if there's an established preference for `string.gsub` or `ngx.re.gsub` (I went with the latter).